File: src/commands/copy.rs
use arboard::Clipboard;
use colored::*;
use inquire::Select;

use crate::vault;

/// Copy a single key's value directly to clipboard — nothing written to disk.
pub fn run() {
    let vault = vault::load();

    if vault.services.is_empty() {
        println!("{}", "  Vault is empty.".yellow());
        return;
    }

    // Step 1: pick the group
    let groups: Vec<String> = vault.services.keys().cloned().collect();
    let group = Select::new("Select group:", groups)
        .prompt()
        .unwrap_or_else(|_| std::process::exit(0));

    let keys_map = vault.services.get(&group).unwrap();

    // Step 2: pick the key
    let key_names: Vec<String> = keys_map.keys().cloned().collect();
    let key = Select::new("Select key to copy:", key_names)
        .prompt()
        .unwrap_or_else(|_| std::process::exit(0));

    let value = keys_map.get(&key).unwrap();

    // Step 3: copy to clipboard
    let mut clipboard = Clipboard::new().unwrap_or_else(|e| {
        eprintln!("Clipboard error: {}", e);
        std::process::exit(1);
    });

    clipboard.set_text(value).unwrap_or_else(|e| {
        eprintln!("Failed to set clipboard: {}", e);
        std::process::exit(1);
    });

    println!(
        "\n  {} {} → {} copied to clipboard.",
        "✓".green().bold(),
        group.cyan(),
        key.cyan().bold()
    );
    println!("  {}", "Value is NOT on disk. Clipboard will clear on reboot.".dimmed());
}

File: ts/src/commands/copy.ts
// Copy a single key's value directly to clipboard — nothing written to disk.
// Ported from rust/src/commands/copy.rs.

import chalk from "chalk";
import clipboard from "clipboardy";
import { select } from "@inquirer/prompts";
import * as vault from "../lib/vault.ts";
import * as terminal from "../lib/terminal.ts";

export async function run(): Promise<void> {
  const data = await vault.load();

  const groups = Object.keys(data.services).sort();
  if (groups.length === 0) {
    console.log(chalk.yellow("  Vault is empty."));
    return;
  }

  let group: string;
  try {
    group = await select({
      message: "Select group:",
      choices: groups.map((g) => ({ value: g, name: g })),
    });
  } catch {
    terminal.cleanupAndExit(0);
  }

  const keysMap = data.services[group] ?? {};
  const keyNames = Object.keys(keysMap).sort();

  let key: string;
  try {
    key = await select({
      message: "Select key to copy:",
      choices: keyNames.map((k) => ({ value: k, name: k })),
    });
  } catch {
    terminal.cleanupAndExit(0);
  }

  const value = keysMap[key];
  if (value === undefined) {
    console.error(chalk.red(`Key '${key}' not found in group '${group}'.`));
    terminal.cleanupAndExit(1);
  }

  try {
    await clipboard.write(value);
  } catch (e) {
    console.error(`Failed to set clipboard: ${(e as Error).message}`);
    terminal.cleanupAndExit(1);
  }

  console.log(
    `\n  ${chalk.green.bold("✓")} ${chalk.cyan(group)} → ${chalk.cyan.bold(key)} copied to clipboard.`,
  );
  console.log(`  ${chalk.dim("Value is NOT on disk. Clipboard will clear on reboot.")}`);
}
